File: app/pipeline.py
"""Orchestrierung: verkettet Erkennung, Ausgleich, Pose, Korrektur, Entzerrung, PDF.

Die Routen in main.py bleiben dadurch reines Transportgeschaeft, und der komplette
Rechenweg ist ohne HTTP testbar.
"""

from __future__ import annotations

import time
from dataclasses import dataclass
from datetime import datetime

import cv2
import numpy as np

from app import config, i18n
from app.notices import AppError, NoticeList
from app.pdf.build import BuildResult, ExportOptions, build_footer_lines, build_pdf
from app.schemas import AdjustRequest, ExportImageRequest, ExportRequest, SolveRequest
from app.session import Session
from app.vision import contour as contour_module
from app.vision import encode
from app.vision import enhance
from app.vision import rectify as rectify_module
from app.vision.camera import CameraPose, resolve_pose
from app.vision.detect import detect_markers, draw_detection
from app.vision.extent import Extent, default_crop, extrapolation_fraction, plane_extent
from app.vision.solve import Solution, solve
from app.vision.thickness import effective_homography


@dataclass(frozen=True)
class ImageResult:
    """Der Zuschnitt als Bilddatei - und was die Oberflaeche darueber sagt."""

    data: bytes
    width: int
    height: int


@dataclass
class SolveResult:
    """Alles, was die UI nach dem Entzerren anzeigt - und was der Export braucht."""

    solution: Solution
    pose: CameraPose
    homography_effective: np.ndarray
    correction_factor: float
    thickness_mm: float
    marker_mm: float
    extent: Extent
    crop: Extent
    preview_px_per_mm: float
    notices: NoticeList
    elapsed_s: float


def run_solve(session: Session, request: SolveRequest) -> SolveResult:
    """Vom Foto zur entzerrten Vorschau samt Qualitaetsbericht."""
    started = time.perf_counter()
    notices = NoticeList()

    if not session.markers:
        session.markers = detect_markers(session.photo.bgr)
        session.write_preview("detected", draw_detection(session.photo.bgr, session.markers))

    solution = solve(
        session.markers, request.marker_mm, request.mode, notices, request.spacing_mm
    )

    pose = resolve_pose(
        solution.homography,
        session.photo.focal35_mm,
        session.photo.width,
        session.photo.height,
        request.thickness_mm,
        request.camera_height_mm,
        notices,
    )
    homography_effective, factor = effective_homography(
        solution.homography, pose, request.thickness_mm
    )

    extent = plane_extent(
        homography_effective, session.photo.width, session.photo.height, solution.hull_mm
    )
    crop = default_crop(extent, solution.hull_mm)

    preview_px_per_mm = rectify_module.preview_px_per_mm(extent)
    preview = rectify_module.rectify(
        session.photo.bgr,
        homography_effective,
        extent,
        preview_px_per_mm,
        source_px_per_mm=solution.px_per_mm,
    )
    session.write_preview("rectified", preview)

    result = SolveResult(
        solution=solution,
        pose=pose,
        homography_effective=homography_effective,
        correction_factor=factor,
        thickness_mm=request.thickness_mm,
        marker_mm=request.marker_mm,
        extent=extent,
        crop=crop,
        preview_px_per_mm=preview_px_per_mm,
        notices=notices,
        elapsed_s=time.perf_counter() - started,
    )
    session.state["solve"] = result
    return result


def run_adjust(session: Session, request: AdjustRequest) -> dict[str, object]:
    """Regler auf die entzerrte Vorschau anwenden - die Antwort fuer den Live-Regler.

    Gearbeitet wird auf dem bereits geschriebenen Vorschau-JPEG, nicht auf einer
    frischen Entzerrung: der Regler soll waehrend des Ziehens antworten, und eine
    Entzerrung des vollen Fotos dauert um Groessenordnungen laenger. Erlaubt ist
    die Abkuerzung, weil die Aufbereitung kosmetisch ist (app/vision/enhance.py) -
    sie taugt am kleinen Bild genau wie am grossen. Verbindlich fuer den Druck ist
    trotzdem allein der Export: der wendet dieselben Regler auf das volle Raster an.
    """
    solved = session.state.get("solve")
    if not isinstance(solved, SolveResult):
        raise AppError("not_solved", "session_id")

    options = request.adjust.to_enhance()
    if options.is_identity:
        # Nichts zu tun: die unveraenderte Vorschau liegt bereits auf der Platte.
        # Eine zweite, Pixel fuer Pixel gleiche Datei waere nur ein Umweg.
        return {"preview": _preview_payload(session, "rectified", solved)}

    rectified = cv2.imread(str(session.preview_path("rectified")))
    if rectified is None:
        raise AppError("preview_missing")

    session.write_preview("adjusted", enhance.adjust(rectified, options))
    return {"preview": _preview_payload(session, "adjusted", solved)}


def _rectified_crop(
    session: Session, request: ExportRequest | ExportImageRequest
) -> tuple[SolveResult, np.ndarray, Extent, float]:
    """Der gemeinsame Anfang beider Exporte: entzerren und aufbereiten.

    `request` muss `crop_mm`, `dpi` und `adjust` tragen - mehr wird hier nicht
    angefasst. Deshalb passen beide Anfragetypen hinein, ohne dass einer den
    anderen erben muesste: was sie unterscheidet (Seitenformat gegen Dateiformat),
    faengt erst NACH dieser Funktion an.
    """
    solved = session.state.get("solve")
    if not isinstance(solved, SolveResult):
        raise AppError("not_solved", "session_id")

    crop = Extent(request.crop_mm.x0, request.crop_mm.y0, request.crop_mm.x1, request.crop_mm.y1)
    if crop.width <= 0.0 or crop.height <= 0.0:
        raise AppError("empty_crop", "crop_mm")

    rectify_module.check_output_budget(crop, request.dpi)
    px_per_mm = rectify_module.px_per_mm_for_dpi(request.dpi)
    rectified = rectify_module.rectify(
        session.photo.bgr,
        solved.homography_effective,
        crop,
        px_per_mm,
        source_px_per_mm=solved.solution.px_per_mm,
    )

    # Aufbereiten NACH dem Entzerren und VOR allem anderen: die Homographie wurde
    # am unberuehrten Foto gemessen, ab hier aendert sich nur noch Farbe und Ton.
    # Die is_identity-Abkuerzung spart bei neutralen Reglern eine vollstaendige
    # Kopie des Rasters - das sind am oberen Ende des Budgets mehrere hundert MB.
    adjust_options = request.adjust.to_enhance()
    if not adjust_options.is_identity:
        rectified = enhance.adjust(rectified, adjust_options)

    return solved, rectified, crop, px_per_mm


def run_export_image(session: Session, request: ExportImageRequest) -> ImageResult:
    """Denselben Zuschnitt als Bilddatei statt als PDF.

    **Was hier fehlt und fehlen muss:** Massstab, Raster, Fusszeile, Schnittmarken,
    Klebeplan. Alles davon ist ein AUFDRUCK auf einem Ausdruck - auf einem Bild
    waere es Bildinhalt, den ein nachgelagertes Programm nicht von der Schablone
    unterscheiden koennte.

    Was NICHT fehlt, ist die Massangabe: die Auflösung steht in der Datei
    (app/vision/encode.py). Ein Pixel ist damit 25,4/dpi Millimeter, und das laesst
    sich lesen statt raten.
    """
    _, rectified, _, _ = _rectified_crop(session, request)
    return ImageResult(
        data=encode.encode_image(rectified, request.image_format, request.dpi),
        width=int(rectified.shape[1]),
        height=int(rectified.shape[0]),
    )


def run_export(session: Session, request: ExportRequest) -> BuildResult:
    """Vom gewaehlten Zuschnitt zum druckfertigen PDF."""
    # Die Sprache des Ausdrucks kommt aus der Anfrage. normalise ist idempotent -
    # das Schema hat den Wert schon abgebildet; hier steht es noch einmal, damit
    # auch ein von Hand gebautes ExportRequest nicht mit "en-GB" durchrutscht.
    locale = i18n.normalise(request.locale)

    solved, rectified, crop, px_per_mm = _rectified_crop(session, request)

    contour_mm = None
    if request.contour:
        # Gesucht wird auf dem AUFBEREITETEN Bild - also auf genau dem, das gleich
        # gedruckt wird. Der Benutzer stellt die Regler ja gerade deshalb, damit
        # eine blasse Bleistiftlinie ueberhaupt als Kante erkennbar wird; auf dem
        # rohen Bild fiele sie durch und die Kontur bliebe leer. Umgekehrt waere
        # eine Kontur aus dem rohen Bild auf dem aufbereiteten Ausdruck eine
        # zweite, andere Wahrheit auf demselben Blatt.
        # Millimeter kostet das nichts: die Aufbereitung faerbt Pixel, sie
        # verschiebt keine - welche Kante gefunden wird, aendert sich, wo sie
        # liegt, nicht (nachgewiesen in tests/test_enhance.py).
        contour_mm = contour_module.find_contour_mm(rectified, px_per_mm)

    options = ExportOptions(
        dpi=request.dpi,
        layout=request.layout,
        page_format=request.page_format,
        orientation=request.orientation,
        overlap_mm=request.overlap_mm,
        printer_margin_mm=request.printer_margin_mm,
        page_margin_mm=request.page_margin_mm,
        show_scalebar=request.overlays.scalebar,
        show_grid=request.overlays.grid,
        show_footer=request.overlays.footer,
        show_marks=request.overlays.marks,
        tile_overview=request.tile_overview,
        contour=request.contour,
        title=i18n.translate(
            "pdf.document_title", locale, width=f"{crop.width:.0f}", height=f"{crop.height:.0f}"
        ),
        locale=locale,
    )

    footer = build_footer_lines(
        _footer_meta(session, solved, crop, request, contour_mm, locale), locale
    )
    return build_pdf(rectified, crop.width, crop.height, options, footer, contour_mm)


def _footer_meta(
    session: Session,
    solved: SolveResult,
    crop: Extent,
    request: ExportRequest,
    contour_mm: np.ndarray | None,
    locale: str = config.DEFAULT_LOCALE,
) -> dict[str, object]:
    """Die Metadaten der Fusszeile - ein Ausdruck soll spaeter nachvollziehbar sein."""
    pose = solved.pose
    if pose.height_mm is None:
        camera = i18n.translate("pdf.footer.camera_unknown", locale)
    else:
        tilt = (
            ""
            if pose.tilt_deg is None
            else i18n.translate("pdf.footer.camera_tilt", locale, tilt_deg=f"{pose.tilt_deg:.0f}")
        )
        camera = i18n.translate(
            "pdf.footer.camera",
            locale,
            height_mm=f"{pose.height_mm:.0f}",
            tilt=tilt,
            source=pose.source,
        )

    object_text = i18n.translate(
        "pdf.footer.object", locale, width=f"{crop.width:.1f}", height=f"{crop.height:.1f}"
    )
    if contour_mm is not None and len(contour_mm) >= 2:
        width_mm, height_mm = contour_module.bounding_box_mm(contour_mm)
        object_text += i18n.translate(
            "pdf.footer.contour", locale, width=f"{width_mm:.1f}", height=f"{height_mm:.1f}"
        )

    # Der Modus IST der Schluessel - fuer jeden gibt es pdf.footer.mode_*.
    mode_key = solved.solution.mode
    return {
        "object_mm": object_text,
        "dpi": request.dpi,
        "scale": i18n.translate(
            "pdf.footer.scale_source", locale, mm_per_px=f"{solved.solution.mm_per_px:.4f}"
        ),
        "mode": i18n.translate(f"pdf.footer.mode_{mode_key}", locale),
        "marker_ids": ",".join(str(f.marker_id) for f in solved.solution.markers),
        "marker_mm": f"{solved.marker_mm:.1f}",
        "rms": f"{solved.solution.rms_px:.2f} px / {solved.solution.rms_mm:.3f} mm",
        "camera": camera,
        "thickness": f"{solved.thickness_mm:.1f} mm (k={solved.correction_factor:.5f})",
        "extrapolation": f"{extrapolation_fraction(crop, solved.solution.hull_mm) * 100:.0f} %",
        "source": session.filename,
        "timestamp": datetime.now().strftime("%Y-%m-%d %H:%M"),
    }


def _preview_payload(session: Session, kind: str, result: SolveResult) -> dict[str, object]:
    """Die Angaben zu einer Vorschau: wo sie liegt und wie sie in mm zu lesen ist.

    Eine Stelle fuer beide Wege - /api/solve und /api/adjust liefern dieselbe Form,
    sodass die Oberflaeche nach dem Regeln nichts anderes auszuwerten hat als nach
    dem Entzerren.
    """
    return {
        "url": f"/api/preview/{session.session_id}/{kind}?t={_cache_buster()}",
        "extent_mm": result.extent.as_dict(),
        "px_per_mm": round(result.preview_px_per_mm, 5),
    }


def _cache_buster() -> int:
    """Ein bei jedem Aufruf neuer Wert fuer die Vorschau-URL.

    Millisekunden, nicht Sekunden: die Vorschau wird unter demselben Dateinamen
    ueberschrieben, und der Regler tut das mehrmals je Sekunde. Mit sekundengenauem
    Stempel bekaeme der Browser zweimal dieselbe URL - und zeigte das alte Bild.
    """
    return time.time_ns() // 1_000_000


def solve_response(
    session: Session, result: SolveResult, locale: str = config.DEFAULT_LOCALE
) -> dict[str, object]:
    """SolveResult in die JSON-Form bringen, die das Frontend erwartet."""
    solution = result.solution
    crop = result.crop
    pose = result.pose

    return {
        "session_id": session.session_id,
        "mode_used": solution.mode,
        "markers": [
            {
                "id": fit.marker_id,
                "corners_px": fit.corners_px.tolist(),
                "side_mm_measured": round(fit.side_mm_measured, 3),
                "residual_px": round(fit.residual_px, 3),
                "rotation_deg": round(fit.rotation_deg, 2),
            }
            for fit in solution.markers
        ],
        "rms_px": round(solution.rms_px, 3),
        "rms_mm": round(solution.rms_mm, 4),
        "mm_per_px": round(solution.mm_per_px, 5),
        "camera": {
            "source": pose.source,
            "focal_px": None if pose.focal_px is None else round(pose.focal_px, 1),
            "height_mm": None if pose.height_mm is None else round(pose.height_mm, 1),
            "nadir_mm": None if pose.nadir_mm is None else [round(v, 1) for v in pose.nadir_mm],
            "tilt_deg": None if pose.tilt_deg is None else round(pose.tilt_deg, 1),
        },
        "thickness_mm": result.thickness_mm,
        "scale_correction_k": round(result.correction_factor, 6),
        "hull_mm": solution.hull_mm.round(3).tolist(),
        "extent_mm": result.extent.as_dict(),
        "default_crop_mm": crop.as_dict(),
        "extrapolation_default": round(
            extrapolation_fraction(crop, solution.hull_mm), 4
        ),
        "preview": {
            **_preview_payload(session, "rectified", result),
            "detected_url": f"/api/preview/{session.session_id}/detected?t={_cache_buster()}",
        },
        "limits": {
            "dpi_choices": list(config.DPI_CHOICES),
            "max_output_mpx": config.MAX_OUTPUT_MPX,
            "extrapolation_warn": config.EXTRAPOLATION_WARN_FRAC,
        },
        "elapsed_s": round(result.elapsed_s, 2),
        "warnings": result.notices.as_dicts(locale),
    }

File: web/vision/pipeline.js
/**
 * pipeline.js - die Rechenkette im Browser. Gegenstueck zu app/pipeline.py.
 *
 * Erkennung, Ausgleich, Pose, Dickenkorrektur, Entzerrung, Aufbereitung, PDF -
 * in derselben Reihenfolge wie dort, mit denselben Zwischenergebnissen. Was auf
 * dem Server eine Sitzung auf der Platte ist, ist hier ein Objekt im Speicher;
 * mehr Unterschied ist es nicht.
 *
 * Die Antwortformen sind ABSICHTLICH die des Servers, Feld fuer Feld. Nur so
 * bleibt app/static/js/ unveraendert - und nur so laesst sich der Ortsbetrieb
 * gegen den Serverbetrieb halten, indem man zwei JSON nebeneinanderlegt.
 */

import * as constants from "../constants.js";
import { buildFooterLines, buildPdf, exportOptions } from "../pdf/build.js";
import { translate } from "../pdf/i18n.js";
import { resolvePose } from "./camera.js";
import { boundingBoxMm, findContourMm } from "./contour.js";
import { withImage } from "./core.js";
import { adjust, adjustOptions, isIdentity } from "./enhance.js";
import {
    defaultCrop,
    extent,
    extrapolationFraction,
    height,
    planeExtent,
    width,
} from "./extent.js";
import { toJpegBytes } from "./image.js";
import { AppError, NoticeList } from "./notices.js";
import {
    checkOutputBudget,
    previewPxPerMm,
    pxPerMmForDpi,
    rectify,
} from "./rectify.js";
import { solve } from "./solve.js";
import { effectiveHomography } from "./thickness.js";

/** Marker im Foto finden. Das Ergebnis ueberlebt die Sitzung - es haengt nur am Foto. */
export function detectMarkers(core, photo) {
    const found = withImage(core, photo.image, (pointer) =>
        core.detectMarkers(
            pointer,
            photo.image.width,
            photo.image.height,
            photo.image.width * photo.image.channels,
            photo.image.channels,
            true,
        ),
    );
    return found.map((marker) => ({
        id: marker.id,
        corners: marker.corners,
        areaPx: core.quadArea(marker.corners),
    }));
}

/** Vom Foto zur entzerrten Vorschau samt Qualitaetsbericht. */
export function runSolve(core, session, request) {
    const started = performance.now();
    const notices = new NoticeList();

    if (session.markers === null) {
        session.markers = detectMarkers(core, session.photo);
    }

    const solution = solve(
        core,
        session.markers,
        request.marker_mm,
        request.mode,
        notices,
        [request.spacing_x_mm, request.spacing_y_mm],
    );

    const pose = resolvePose(
        core,
        solution.homography,
        session.photo.focal35Mm,
        session.photo.image.width,
        session.photo.image.height,
        request.thickness_mm,
        request.camera_height_mm,
        notices,
    );
    const corrected = effectiveHomography(solution.homography, pose, request.thickness_mm);

    const area = planeExtent(
        core,
        corrected.homography,
        session.photo.image.width,
        session.photo.image.height,
        solution.hullMm,
    );
    const crop = defaultCrop(area, solution.hullMm);

    const pxPerMm = previewPxPerMm(area);
    const preview = rectify(
        core,
        session.photo.image,
        corrected.homography,
        area,
        pxPerMm,
        solution.pxPerMm,
    );

    const result = {
        solution,
        pose,
        homographyEffective: corrected.homography,
        correctionFactor: corrected.factor,
        thicknessMm: request.thickness_mm,
        markerMm: request.marker_mm,
        extent: area,
        crop,
        previewPxPerMm: pxPerMm,
        preview,
        adjusted: null,
        notices,
        elapsedS: (performance.now() - started) / 1000.0,
    };
    session.solve = result;
    return result;
}

/**
 * Regler auf die entzerrte Vorschau anwenden - die Antwort fuer den Live-Regler.
 *
 * Gearbeitet wird auf dem bereits berechneten Vorschauraster, nicht auf einer
 * frischen Entzerrung: der Regler soll waehrend des Ziehens antworten, und eine
 * Entzerrung des vollen Fotos dauert um Groessenordnungen laenger. Erlaubt ist
 * die Abkuerzung, weil die Aufbereitung kosmetisch ist - sie taugt am kleinen
 * Bild genau wie am grossen. Verbindlich fuer den Druck ist trotzdem allein der
 * Export: der wendet dieselben Regler auf das volle Raster an.
 */
export function runAdjust(core, session, request) {
    const solved = session.solve;
    if (solved === null) throw new AppError("not_solved", "session_id");

    const options = adjustOptions(request.adjust);
    if (isIdentity(core, options)) {
        solved.adjusted = null;
        return { kind: "rectified", raster: solved.preview };
    }

    solved.adjusted = adjust(core, solved.preview, options);
    return { kind: "adjusted", raster: solved.adjusted };
}

/** Vom gewaehlten Zuschnitt zum druckfertigen PDF. */
export async function runExport(core, session, request) {
    const solved = session.solve;
    if (solved === null) throw new AppError("not_solved", "session_id");

    const locale = normaliseLocale(request.locale);
    const crop = extent(
        request.crop_mm.x0,
        request.crop_mm.y0,
        request.crop_mm.x1,
        request.crop_mm.y1,
    );
    if (width(crop) <= 0.0 || height(crop) <= 0.0) throw new AppError("empty_crop", "crop_mm");

    const pxPerMm = pxPerMmForDpi(request.dpi);
    // Aufbereiten NACH dem Entzerren und VOR allem anderen: die Homographie wurde
    // am unberuehrten Foto gemessen, ab hier aendert sich nur noch Farbe und Ton.
    const options = adjustOptions(request.adjust);

    /** Entzerren, aufbereiten, kodieren - fuer ein Rechteck in Zuschnitt-Millimetern. */
    const render = async (area) => {
        // Dieselbe Pruefung wie bisher, nur auf dem, was wirklich belegt wird.
        // Blattweise ist das ein A4-Blatt und geht ueberall durch; am Stueck ist
        // es der ganze Zuschnitt, und dann soll dieser Abbruch kommen und nicht
        // ein OutOfMemoryError im Entzerren.
        checkOutputBudget(core, area, request.dpi);
        let raster = rectify(
            core,
            session.photo.image,
            solved.homographyEffective,
            area,
            pxPerMm,
            solved.solution.pxPerMm,
        );
        if (!isIdentity(core, options)) {
            raster = adjust(core, raster, options);
        }
        return { raster, jpeg: await toJpegBytes(raster) };
    };

    // Am Stueck oder blattweise - und warum, steht in needsWholeRaster.
    let source;
    let contourMm = null;
    if (needsWholeRaster(request, options)) {
        const whole = await render(crop);
        // Gesucht wird auf dem AUFBEREITETEN Bild - also auf genau dem, das gleich
        // gedruckt wird. Millimeter kostet das nichts: die Aufbereitung faerbt
        // Pixel, sie verschiebt keine.
        contourMm = request.contour ? findContourMm(core, whole.raster, pxPerMm) : null;
        source = whole.jpeg;
    } else {
        // Die Bildquelle fuer web/pdf/build.js: ein Blatt, ein Raster. Der
        // Spitzenbedarf haengt damit am Blatt und nicht mehr am Zuschnitt - genau
        // daran ist der Export auf dem Telefon gescheitert.
        source = async (xMm, yMm, widthMm, heightMm) =>
            (await render(extent(
                crop.x0 + xMm,
                crop.y0 + yMm,
                crop.x0 + xMm + widthMm,
                crop.y0 + yMm + heightMm,
            ))).jpeg;
    }

    const options_ = exportOptions({
        dpi: request.dpi,
        layout: request.layout,
        pageFormat: request.page_format,
        orientation: request.orientation,
        overlapMm: request.overlap_mm,
        printerMarginMm: request.printer_margin_mm,
        pageMarginMm: request.page_margin_mm,
        showScalebar: request.overlays.scalebar,
        showGrid: request.overlays.grid,
        showFooter: request.overlays.footer,
        showMarks: request.overlays.marks,
        tileOverview: request.tile_overview,
        contour: request.contour,
        title: translate("pdf.document_title", locale, {
            width: width(crop).toFixed(0),
            height: height(crop).toFixed(0),
        }),
        locale,
    });

    const footer = buildFooterLines(
        footerMeta(core, session, solved, crop, request, contourMm, locale),
        locale,
    );
    return buildPdf(source, width(crop), height(crop), options_, footer, contourMm);
}

/**
 * Braucht dieser Export das ganze Raster auf einmal?
 *
 * <b>Der Regelfall ist nein</b>, und das ist der Unterschied zwischen einem
 * Export, der auf einem Telefon durchlaeuft, und einem, der es nicht tut: ein
 * Zuschnitt von 810 x 1153 mm sind bei 300 dpi 130 Megapixel und 373 MiB an
 * einem Stueck, ein A4-Blatt derselben Aufloesung sind 26 MB.
 *
 * Blattweise ist das Ergebnis Bit fuer Bit dasselbe, solange nichts ueber
 * Pixelgrenzen hinweg rechnet. Drei Dinge tun das:
 *
 *  - <b>Der Umriss</b> wird auf dem fertigen Bild gesucht. Blattweise faende er
 *    je Blatt einen eigenen und nirgends den ganzen.
 *  - <b>Lokaler Kontrast</b> ist CLAHE: Histogramme ueber ein Gitter, das ueber
 *    das GANZE Bild gelegt wird. Blattweise bekaeme jedes Blatt sein eigenes.
 *  - <b>Kantenschaerfe und Kantenzeichnung</b> greifen in die Nachbarschaft
 *    (Unschaerfemaske bzw. Canny). An der Blattkante fehlt die.
 *
 * Alle drei stehen per Vorgabe aus. Die uebrigen Regler - Helligkeit, Kontrast,
 * Saettigung, Graustufen, Umkehr, Farbbetonung, Schwelle - rechnen Pixel fuer
 * Pixel; sie sind blattweise exakt dasselbe und stehen deshalb nicht hier.
 *
 * Bei einer Einzelseite ist die Frage ohnehin gegenstandslos: dort gibt es nur
 * ein Blatt, und web/pdf/build.js fragt die Quelle dann einmal nach allem.
 */
function needsWholeRaster(request, options) {
    return (
        Boolean(request.contour) ||
        options.local_contrast > 0.0 ||
        options.edge_boost > 0.0 ||
        options.edge_overlay > 0.0
    );
}

/** Die Metadaten der Fusszeile - ein Ausdruck soll spaeter nachvollziehbar sein. */
function footerMeta(core, session, solved, crop, request, contourMm, locale) {
    const pose = solved.pose;
    let camera;
    if (pose.heightMm === null) {
        camera = translate("pdf.footer.camera_unknown", locale);
    } else {
        const tilt =
            pose.tiltDeg === null
                ? ""
                : translate("pdf.footer.camera_tilt", locale, { tilt_deg: pose.tiltDeg.toFixed(0) });
        camera = translate("pdf.footer.camera", locale, {
            height_mm: pose.heightMm.toFixed(0),
            tilt,
            source: pose.source,
        });
    }

    let objectText = translate("pdf.footer.object", locale, {
        width: width(crop).toFixed(1),
        height: height(crop).toFixed(1),
    });
    if (contourMm !== null && contourMm.length >= 2) {
        const [contourWidth, contourHeight] = boundingBoxMm(contourMm);
        objectText += translate("pdf.footer.contour", locale, {
            width: contourWidth.toFixed(1),
            height: contourHeight.toFixed(1),
        });
    }

    // Der Modus IST der Schluessel - fuer jeden gibt es pdf.footer.mode_*.
    const modeKey = solved.solution.mode;
    return {
        object_mm: objectText,
        dpi: request.dpi,
        scale: translate("pdf.footer.scale_source", locale, {
            mm_per_px: solved.solution.mmPerPx.toFixed(4),
        }),
        mode: translate(`pdf.footer.mode_${modeKey}`, locale),
        marker_ids: solved.solution.markers.map((fit) => fit.id).join(","),
        marker_mm: solved.markerMm.toFixed(1),
        rms: `${solved.solution.rmsPx.toFixed(2)} px / ${solved.solution.rmsMm.toFixed(3)} mm`,
        camera,
        thickness: `${solved.thicknessMm.toFixed(1)} mm (k=${solved.correctionFactor.toFixed(5)})`,
        extrapolation: `${(extrapolationFraction(core, crop, solved.solution.hullMm) * 100).toFixed(0)} %`,
        source: session.filename,
        timestamp: timestamp(),
    };
}

/** "2026-09-08 03:47" in der Zeitzone des Geraets - wie datetime.now() am Desktop. */
function timestamp() {
    const now = new Date();
    const pad = (value) => String(value).padStart(2, "0");
    return (
        `${now.getFullYear()}-${pad(now.getMonth() + 1)}-${pad(now.getDate())} ` +
        `${pad(now.getHours())}:${pad(now.getMinutes())}`
    );
}

function normaliseLocale(value) {
    if (!value) return constants.DEFAULT_LOCALE;
    const primary = String(value).trim().toLowerCase().replace(/_/g, "-").split("-")[0];
    return constants.SUPPORTED_LOCALES.includes(primary) ? primary : constants.DEFAULT_LOCALE;
}

/** Die JSON-Form, die app/static/js/report.js und crop-rect.js erwarten. */
export function solveResponse(session, result, core, previewUrl, detectedUrl, describe) {
    const solution = result.solution;
    const pose = result.pose;

    return {
        session_id: session.id,
        mode_used: solution.mode,
        markers: solution.markers.map((fit) => ({
            id: fit.id,
            corners_px: pairs(fit.cornersPx),
            side_mm_measured: round(fit.sideMmMeasured, 3),
            residual_px: round(fit.residualPx, 3),
            rotation_deg: round(fit.rotationDeg, 2),
        })),
        rms_px: round(solution.rmsPx, 3),
        rms_mm: round(solution.rmsMm, 4),
        mm_per_px: round(solution.mmPerPx, 5),
        camera: {
            source: pose.source,
            focal_px: pose.focalPx === null ? null : round(pose.focalPx, 1),
            height_mm: pose.heightMm === null ? null : round(pose.heightMm, 1),
            nadir_mm: pose.nadirMm === null ? null : pose.nadirMm.map((value) => round(value, 1)),
            tilt_deg: pose.tiltDeg === null ? null : round(pose.tiltDeg, 1),
        },
        thickness_mm: result.thicknessMm,
        scale_correction_k: round(result.correctionFactor, 6),
        hull_mm: pairs(solution.hullMm).map(([x, y]) => [round(x, 3), round(y, 3)]),
        extent_mm: { ...result.extent },
        default_crop_mm: { ...result.crop },
        extrapolation_default: round(
            extrapolationFraction(core, result.crop, solution.hullMm),
            4,
        ),
        preview: {
            url: previewUrl,
            extent_mm: { ...result.extent },
            px_per_mm: round(result.previewPxPerMm, 5),
            detected_url: detectedUrl,
        },
        limits: {
            dpi_choices: [...constants.DPI_CHOICES],
            max_output_mpx: constants.outputBudgetMpx(),
            extrapolation_warn: constants.EXTRAPOLATION_WARN_FRAC,
        },
        elapsed_s: round(result.elapsedS, 2),
        warnings: result.notices.asDicts(describe),
    };
}

function pairs(flat) {
    const points = [];
    for (let index = 0; index < flat.length; index += 2) {
        points.push([flat[index], flat[index + 1]]);
    }
    return points;
}

/** Runden wie Pythons round(x, n) - die Oberflaeche zeigt diese Zahlen an. */
function round(value, digits) {
    const factor = 10 ** digits;
    return Math.round(value * factor) / factor;
}
